pkg/providers/instancetype/offering/provider.go

```diff
@@ -0,0 +1,274 @@
+			default:
+				panic(fmt.Sprintf("invalid capacity type %q in requirements for instance type %q", capacityType, it.Name))
+			}
+			offering := cloudprovider.Offering{
+				Requirements: scheduling.NewRequirements(
+					scheduling.NewRequirement(karpv1.CapacityTypeLabelKey, corev1.NodeSelectorOpIn, capacityType),
+					scheduling.NewRequirement(corev1.LabelTopologyZone, corev1.NodeSelectorOpIn, zone),
+					scheduling.NewRequirement(v1.LabelCapacityReservationID, corev1.NodeSelectorOpDoesNotExist),
+				),
+				Price:     price,
+				Available: !isUnavailable && hasPrice && itZones.Has(zone) && hasSubnetZone,
+			}
+			if id, ok := subnetZones[zone]; ok {
+				offering.Requirements.Add(scheduling.NewRequirement(v1.LabelTopologyZoneID, corev1.NodeSelectorOpIn, id))
+			}
+			offerings = append(offerings, offering)
+		}
+	}
+
+	for i := range nodeClass.Status.CapacityReservations {
+		if nodeClass.Status.CapacityReservations[i].InstanceType != it.Name {
+			continue
+		}
+		reservation := &nodeClass.Status.CapacityReservations[i]
+
+		isUnavailable := p.unavailableOfferings.IsReservationUnavailable(reservation.ID)
+		_, hasSubnetZone := subnetZones[reservation.AvailabilityZone]
+		price := 0.0
+		if odPrice, ok := p.pricingProvider.OnDemandPrice(ec2types.InstanceType(it.Name)); ok {
+			// Divide the on-demand price by a sufficiently large constant. This allows us to treat the reservation as "free",
+			// while maintaining relative ordering for consolidation. If the pricing details are unavailable for whatever reason,
+			// still succeed to create the offering and leave the price at zero. This will break consolidation, but will allow
+			// users to utilize the instances they're already paying for.
+			price = odPrice / 100_000.0
+		}
+		offering := cloudprovider.Offering{
+			Requirements: scheduling.NewRequirements(
+				scheduling.NewRequirement(karpv1.CapacityTypeLabelKey, corev1.NodeSelectorOpIn, karpv1.CapacityTypeReserved),
+				scheduling.NewRequirement(corev1.LabelTopologyZone, corev1.NodeSelectorOpIn, reservation.AvailabilityZone),
+				scheduling.NewRequirement(v1.LabelCapacityReservationID, corev1.NodeSelectorOpIn, reservation.ID),
+			),
+			Price:     price,
+			Available: !isUnavailable && itZones.Has(reservation.AvailabilityZone) && hasSubnetZone,
+			// TODO: Retrieve available instance count from an in-memory store accounting for launched and terminated instances
+			// in real time, rather than dealing with eventual consistency delays with the NodeClass status.
+			ReservationManager: p.createReservationManager(reservation.ID, reservation.AvailableInstanceCount, snapshotUUID),
+		}
+		if id, ok := subnetZones[reservation.AvailabilityZone]; ok {
+			offering.Requirements.Add(scheduling.NewRequirement(v1.LabelTopologyZoneID, corev1.NodeSelectorOpIn, id))
+		}
+		offerings = append(offerings, offering)
+	}
+	return cloudprovider.Offerings(offerings)
```

Comment:
nit: I don't think you need to do this conversion if you just declare the cloudprovider.Offerings as this type at the top

Reply:
Don't even need to that it turns out, it infers the type just fine